In [1]:
import numpy as np
import pandas as pd

import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv('housing.csv')

In [9]:
data.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [10]:
cols = ['latitude',
'longitude',
'housing_median_age',
'total_rooms',
'total_bedrooms',
'population',
'households',
'median_income',
'median_house_value']

In [11]:
df = data[cols]

In [12]:
df.head()

latitude  longitude  housing_median_age  total_rooms  total_bedrooms  \
0     37.88    -122.23                41.0        880.0           129.0   
1     37.86    -122.22                21.0       7099.0          1106.0   
2     37.85    -122.24                52.0       1467.0           190.0   
3     37.85    -122.25                52.0       1274.0           235.0   
4     37.85    -122.25                52.0       1627.0           280.0   

   population  households  median_income  median_house_value  
0       322.0       126.0         8.3252            452600.0  
1      2401.0      1138.0         8.3014            358500.0  
2       496.0       177.0         7.2574            352100.0  
3       558.0       219.0         5.6431            341300.0  
4       565.0       259.0         3.8462            342200.0

In [13]:
# Question 1
df.isnull().sum()

latitude                0
longitude               0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
dtype: int64

the answer is that total bedrooms has 207 missing values

In [15]:
# Question 2
# What's the median (50% percentile) for variable 'population'?
df['population'].describe()

count    20640.000000
mean      1425.476744
std       1132.462122
min          3.000000
25%        787.000000
50%       1166.000000
75%       1725.000000
max      35682.000000
Name: population, dtype: float64

answer is 1166

Shuffle the initial dataset, use seed 42.
Split your data in train/val/test sets, with 60%/20%/20% distribution.
Make sure that the target value ('median_house_value') is not in your dataframe.
Apply the log transformation to the median_house_value variable using the np.log1p() function.

In [16]:
# breaking the price off for later
price = df.median_house_value

In [17]:
# dropping price from the df
df = df.drop('median_house_value', axis=1)
df.head()

latitude  longitude  housing_median_age  total_rooms  total_bedrooms  \
0     37.88    -122.23                41.0        880.0           129.0   
1     37.86    -122.22                21.0       7099.0          1106.0   
2     37.85    -122.24                52.0       1467.0           190.0   
3     37.85    -122.25                52.0       1274.0           235.0   
4     37.85    -122.25                52.0       1627.0           280.0   

   population  households  median_income  
0       322.0       126.0         8.3252  
1      2401.0      1138.0         8.3014  
2       496.0       177.0         7.2574  
3       558.0       219.0         5.6431  
4       565.0       259.0         3.8462

In [18]:
# getting the length of df making a a  range and shuffling it
n = len(df)
n_train = int(n * 0.6)
n_val = int(n * 0.2)
n_test = n - n_train - n_val
n_train, n_val, n_test

(12384, 4128, 4128)

In [19]:
n, n_train + n_val + n_test

(20640, 20640)

In [20]:
np.random.seed(42)

In [21]:
idx = np.arange(n)
np.random.shuffle(idx)
idx

array([20046,  3024, 15663, ...,  5390,   860, 15795])

In [22]:
df_train = df.iloc[idx[:n_train]]
df_val = df.iloc[idx[n_train:n_train + n_val]]
df_test = df.iloc[idx[n_train + n_val:]]
len(df), len(df_train) + len(df_val) + len(df_test)

(20640, 20640)

In [24]:
# converting price to log price
y = np.log1p(price)
y

0        13.022766
1        12.789687
2        12.771673
3        12.740520
4        12.743154
           ...    
20635    11.265758
20636    11.252872
20637    11.432810
20638    11.346883
20639    11.400887
Name: median_house_value, Length: 20640, dtype: float64

In [26]:
y_train = y.iloc[idx[:n_train]]
y_val = y.iloc[idx[n_train:n_train + n_val]]
y_test = y.iloc[idx[n_train + n_val:]]
len(df), len(df_train) + len(df_val) + len(df_test)

(20640, 20640)

In [27]:
df_train.head()

latitude  longitude  housing_median_age  total_rooms  total_bedrooms  \
20046     36.06    -119.01                25.0       1505.0             NaN   
3024      35.14    -119.46                30.0       2943.0             NaN   
15663     37.80    -122.44                52.0       3830.0             NaN   
20484     34.28    -118.72                17.0       3051.0             NaN   
9814      36.62    -121.93                34.0       2351.0             NaN   

       population  households  median_income  
20046      1392.0       359.0         1.6812  
3024       1565.0       584.0         2.5313  
15663      1310.0       963.0         3.4801  
20484      1705.0       495.0         5.7376  
9814       1063.0       428.0         3.7250

In [28]:
y_train.head()

20046    10.772708
3024     10.732061
15663    13.122367
20484    12.295003
9814     12.535380
Name: median_house_value, dtype: float64

In [29]:
def prepare_X(df, fill_strategy=0):
    
    df = df.copy()
    
    if fill_strategy == 'mean':
        df = df.fillna(df_train.mean())
        
    else:
        df = df.fillna(0)
        
    X = df.values
    
    return X

In [30]:
def train_linear_regression(X, y):
    ones = np.ones(X.shape[0])
    
    X = np.column_stack([ones, X])
    
    XTX = X.T.dot(X)
    XTX_inv = np.linalg.inv(XTX)
    w_full = XTX_inv.dot(X.T).dot(y)
    
    return w_full[0], w_full[1:]

In [31]:
def rmse(y, y_pred):
    error = y - y_pred
    se = error ** 2
    mse = se.mean()
    rmse = np.sqrt(mse)
    
    return rmse

In [ ]:
# Question 3
# with 

In [33]:
# with fillna(0) option
X_train = prepare_X(df_train)

w0, w = train_linear_regression(X_train, y_train)

X_val = prepare_X(df_val)
y_pred = w0 + X_val.dot(w)

round(rmse(y_val, y_pred), 2)

0.33

In [34]:
# with fillna('mean') option
X_train = prepare_X(df_train, fill_strategy='mean')

w0, w = train_linear_regression(X_train, y_train)

X_val = prepare_X(df_val)
y_pred = w0 + X_val.dot(w)

round(rmse(y_val, y_pred), 2)

0.33

so the answer is that they come to the same answer

In [35]:
# Question 4
# possible r values [0, 0.000001, 0.0001, 0.001, 0.01, 0.1, 1, 5, 10]

def train_linear_regression_reg(X, y, r=0.001):
    ones = np.ones(X.shape[0])
    
    X = np.column_stack([ones, X])
    
    XTX = X.T.dot(X)
    XTX = XTX + r * np.eye(XTX.shape[0])
    XTX_inv = np.linalg.inv(XTX)
    w_full = XTX_inv.dot(X.T).dot(y)
    
    return w_full[0], w_full[1:]

In [36]:
for r  in [0, 0.000001, 0.0001, 0.001, 0.01, 0.1, 1, 5, 10]:
    X_train = prepare_X(df_train)
    w0, w = train_linear_regression_reg(X_train, y_train, r=r)

    X_val = prepare_X(df_val)
    y_pred = w0 + X_val.dot(w)
    score = rmse(y_val, y_pred)
    
    print(r, w0, round(score, 2))

0 -11.686975241713926 0.33
1e-06 -11.68695917553698 0.33
0.0001 -11.685368865728286 0.33
0.001 -11.67093131795569 0.33
0.01 -11.52849358570942 0.33
0.1 -10.274500282064466 0.33
1 -4.920480897806385 0.33
5 -1.4820957456166362 0.34
10 -0.7899311832497747 0.34


so the answer would be 0 is the best r, basically to leave it unregularized

In [39]:
scores = []
for s in [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]:
    np.random.seed(s)
    idx = np.arange(n)
    np.random.shuffle(idx)
    
    df_train = df.iloc[idx[:n_train]]
    df_val = df.iloc[idx[n_train:n_train+n_val]]
    df_test = df.iloc[idx[n_train + n_val:]]
    
    y_train = y.iloc[idx[:n_train]]
    y_val = y.iloc[idx[n_train:n_train+n_val]]
    y_test = y.iloc[idx[n_train + n_val:]]
    
    X_train = prepare_X(df_train)

    w0, w = train_linear_regression(X_train, y_train)

    X_val = prepare_X(df_val)
    y_pred = w0 + X_val.dot(w)
    
    
    score = round(rmse(y_val, y_pred), 2)

    
    print(f'rmse for random seed {s}: {score}')
    scores.append(score)
    
print(f'std of the rmse scores: {round(np.std(scores), 3)}')

rmse for random seed 0: 0.34
rmse for random seed 1: 0.34
rmse for random seed 2: 0.33
rmse for random seed 3: 0.34
rmse for random seed 4: 0.34
rmse for random seed 5: 0.34
rmse for random seed 6: 0.35
rmse for random seed 7: 0.34
rmse for random seed 8: 0.35
rmse for random seed 9: 0.34
std of the rmse scores: 0.005


answer to question 5 is 0.005

In [40]:
# Question 6

np.random.seed(9)
idx = np.arange(n)
np.random.shuffle(idx)

df_train = df.iloc[idx[:n_train+n_val]]
df_test = df.iloc[idx[n_train + n_val:]]

y_train = y.iloc[idx[:n_train+n_val]]
y_test = y.iloc[idx[n_train + n_val:]]

X_train = prepare_X(df_train)
w0, w = train_linear_regression_reg(X_train, y_train, r=0.001)

X_test = prepare_X(df_test)
y_pred = w0 + X_test.dot(w)
score = rmse(y_test, y_pred)

print(w0, round(score, 2))

-11.407983344014404 0.35


answer to question 6 is 0.35